### DISCUSSION TITLE
Shutdown script

### DISCUSSION THREAD

**Discussion by GeosDroid** [autokey/autokey#716] [General]
I just want to be able to shutdown or restart with a key combination. Can someone point me to a script to use? Thanks.

**Comment by josephj11** [comment]
There are a couple of parts to this, but both are relatively easy.

First, write two shell script (bash, zsh, ...) that contains the following:

    #!/bin/bash
    ## my_shutdown
    sudo shutdown -h now

and

    #!/bin/bash
    ## my_restart
    sudo shutdown -r now

And save them in $HOME/bin on your account and make them executable.
Then write two AutoKey scripts

    import subprocess
    subprocess.call(["my_shutdown"])

and

    import subprocess
    subprocess.call(["my_restart"])

and assign them to hotkeys of your choice.

That's the core of it, but there's probably more you need to do.
If the `shutdown` command works on your system without sudo, then you can eliminate the `sudo`s and you're done.

If it requires sudo, then you have to get around that because sudo needs to interact with a terminal and you usually don't have one in focus when you press a hotkey.

There are lots of ways to deal with this.

One way is to add `my_shutdown` and `my_restart` to the sudoers file as programs that do not require passwords to run.
This makes life easy (once you figure out the sudoers syntax), but may be a bit of a security risk unless you store those files in a more restricted directory such as `/usr/local/bin` and make sure AutoKey can find them.

Another way would be to add some fancy GUI dialog code to either your bash or AutoKey scripts to ask for your password when they run. Sudo has a `--askpass` option to run a program to ask for your password, but you'd have to look into that. Apparently `ssh-askpass` does this and is installed on my system. I don't believe I have ever explicitly called it.

**Comment by Elliria** [comment]
And yet another way of doing it would be to write a reboot script and a shutdown script, neither of which requires sudo, and use whichever triggers you like for them.

**The reboot script:**
```
import subprocess
subprocess.call(["systemctl","reboot"])
```

**The shut-down script:**
```
import subprocess
subprocess.call(["systemctl","poweroff"])
```

**Reply by josephj11** [reply]
@Elliria That's a cleaner solution.

@GeosDroid Of course, doing a shutdown of a running system (using any method) may have unforeseen consequences if any processes are in  the middle of things like disk writes or database updates. Linux will try to complete some of them and most file systems use journalling to recover upon the next boot, but they're not perfect. So, such an option should be used sparingly.

A less drastic option would be doing a user logout or resetting the X server (which also causes a user logout), but lets the underlying system keep running. This would be a bit more complicated to get right because, at least on my system, KDE will ask me about unsaved files and fail to reset if I don't resolve them first. I have Ctrl+Alt+Backspace configured to reset X. It's a (KDE?) option. Sometimes it screws things up enough that I have to reboot anyway after using it.
